### PR TITLE
feat: Add basic RPC protocol and websocket based server

### PR DESCRIPTION
**Pull request by elkowar**
This PR adds tons of dependencies, and happens to also add an RPC protocol and a small prototype server that is served through websocket.

### CODE REVIEW

File: crates/bazed-core/src/app.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+use futures::StreamExt;
+use tokio::sync::RwLock;
+
+use crate::{document::Document, marked_rope::MarkerId};
+
+pub struct App {
+    document: Option<Document>,
+    event_send: ClientSendHandle,
+    cursors: Vec<MarkerId>,
+}
```
Comment (gwimm) [inline crates/bazed-core/src/app.rs:14]: Please find a more descriptive name for your internal state struct.

Comment (gwimm) [reply crates/bazed-core/src/app.rs:14]: I disagree with "App is the perfect name for the inter-
I also disagree with convention.

File: crates/bazed-core/src/app.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            ToBackend::MouseInput { line, column } => {
+                println!("Lmao, mouse input: {column},{line}")
+            },
+        }
+        Ok(())
+    }
+}
+
+pub async fn start(addr: &str) -> Result<()> {
+    let (send, mut recv) = bazed_rpc::server::wait_for_client(addr).await?;
```
Comment (gwimm) [inline crates/bazed-core/src/app.rs:57]: All of this hinges on a single client?

File: crates/bazed-core/src/lib.rs [hunk 3]
```diff
@@ -1,51 +1,5 @@
 #![forbid(unreachable_pub)]
-use std::{fs::File, io::BufReader, path::PathBuf};
-
-use color_eyre::Result;
-use marked_rope::{MarkedRope, MarkerId};
-use ropey::Rope;
 
+pub mod app;
```
Comment (gwimm) [inline crates/bazed-core/src/lib.rs:3]: My previous argument applies.

File: crates/bazed-rpc/src/keycode.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            Self::U => 'u',
+            Self::V => 'v',
+            Self::W => 'w',
+            Self::X => 'x',
+            Self::Y => 'y',
+            Self::Z => 'z',
+            _ => return None,
+        })
+    }
+}
```
Comment (gwimm) [inline crates/bazed-rpc/src/keycode.rs:130]: Still kinda find it funky that the backend handles raw key input, but let's see :D

File: crates/bazed/src/main.rs [hunk 5]
```diff
@@ -1,3 +1,10 @@
-fn main() {
-    panic!("hello world");
```
Comment (gwimm) [inline crates/bazed/src/main.rs:2]: Sad to see it go.

**Review by gwimm — APPROVED** [review]
Good stuff don't really see anything i vehemently disagree with.

File: Cargo.toml [hunk 6]
```diff
@@ -16,9 +16,18 @@ license = "Apache-2.0 OR MIT"
...
 thiserror = "1.0.37"
 
+serde_json = "1.0.89"
+serde = { version = "1.0.151", features = ["derive"] }
+uuid = { version = "1.2.2", features = ["v4", "serde"] }
+
+futures = "0.3.25"
+tokio = { version = "1.23.0", features = ["rt", "net", "macros", "time", "sync"] }
+
+bazed-rpc = { path = "crates/bazed-rpc" }
```
Comment (buffet) [inline Cargo.toml]: These need versions

File: crates/bazed-core/src/app.rs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                    self.event_send
+                        .send_rpc_notification(ToFrontend::UpdateText {
+                            id: document.id().0.to_owned(),
+                            text: document.content_to_string(),
+                        })
+                        .await?;
+                }
+            },
+            ToBackend::MouseInput { line, column } => {
+                println!("Lmao, mouse input: {column},{line}")
```
Comment (buffet) [inline crates/bazed-core/src/app.rs:49]: Use proper logging pls.

Comment (gwimm) [inline crates/bazed-core/src/app.rs:49]: Use better message.

**Review by gwimm — CHANGES_REQUESTED** [review]

File: crates/bazed-core/src/app.rs (hunk 7 above)
Comment (elkowar) [reply crates/bazed-core/src/app.rs:49]: no u

File: crates/bazed-core/src/document.rs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        let rope = MarkedRope::new(Rope::from_reader(BufReader::new(File::open(&path)?))?);
+        Ok(Self {
+            id: DocumentId::gen(),
+            source: Some(path),
+            dirty: false,
+            content: rope,
+        })
+    }
+
+    pub fn source(&self) -> Option<&PathBuf> {
```
Comment (buffet) [inline crates/bazed-core/src/document.rs:34]: `source` actually kinda weird, `file` would be better.

File: crates/bazed-core/src/document.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    pub fn delete_char(&mut self, m: MarkerId) -> Result<(), crate::marked_rope::Error> {
+        let res = self.content.remove_char(m);
+        if res.is_ok() {
+            self.dirty = true;
+        }
+        res
+    }
+
+    pub fn sticky_marker_start(&mut self) -> MarkerId {
```
Comment (buffet) [inline crates/bazed-core/src/document.rs:58]: This sounds like it starts a sticky marker.

File: crates/bazed-rpc/src/keycode.rs [hunk 10]
```diff
@@ -0,0 +1,130 @@
+use serde::{Deserialize, Serialize};
+
+/// A combination of held [Modifier]s and a [Key].
+#[derive(Debug, Serialize, Deserialize)]
+pub struct KeyInput {
+    pub modifiers: Vec<Modifier>,
+    pub key: Key,
+}
+
+/// modifier keys
```
Comment (buffet) [inline crates/bazed-rpc/src/keycode.rs:10]: Please use a spellchecker.

File: crates/bazed-rpc/src/keycode.rs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+pub enum Modifier {
+    Ctrl,
+    Alt,
+    Shift,
+    Win,
+}
+
+/// Any relevant letter, symbol of nav-key on a standard qwerty keyboard.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Key {
```
Comment (buffet) [inline crates/bazed-rpc/src/keycode.rs:21]: eww, yuck

File: crates/bazed-rpc/src/server.rs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    let server_listener = tokio::net::TcpListener::bind(addr)
+        .await
+        .context("Failed to start tcp server")?;
+
+    // for now, we only accept a single client. This will need to be a loop later.
+    let (stream, _) = server_listener.accept().await?;
+    let ws_stream = tokio_tungstenite::accept_async(stream).await?;
+    let (ws_send, mut ws_recv) = ws_stream.split();
+
+    let (mut to_backend_send, to_backend_recv) = futures::channel::mpsc::channel::<ToBackend>(10);
```
Comment (buffet) [inline crates/bazed-rpc/src/server.rs]: Why 10?

File: crates/bazed-rpc/src/server.rs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    tokio::spawn(async move {
+        while let Some(msg) = ws_recv.next().await {
+            match msg {
+                Ok(tungstenite::Message::Text(json)) => {
+                    match serde_json::from_str::<ToBackend>(&json) {
+                        Ok(x) => {
+                            if let Err(err) = to_backend_send.try_send(x) {
+                                eprintln!("Stopping websocket receiver forwarding loop: {err}");
```
Comment (buffet) [inline crates/bazed-rpc/src/server.rs:43]: Please use proper logging.

File: crates/bazed-rpc/src/server.rs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+                                break;
+                            }
+                        },
+                        Err(err) => eprintln!("Error parsing rpc message: {err:?}"),
```
Comment (buffet) [inline crates/bazed-rpc/src/server.rs:47]: Please use proper logging.

File: crates/bazed-rpc/src/server.rs [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+                    }
+                },
+                Ok(other) => println!("Got unsupported websocket message: {other:?}"),
```
Comment (buffet) [inline crates/bazed-rpc/src/server.rs:50]: Please use proper logging.

File: crates/bazed-rpc/src/server.rs [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+                Err(err) => eprintln!("Got error message from websocket: {err:?}"),
```
Comment (buffet) [inline crates/bazed-rpc/src/server.rs:51]: Please use proper logging.

File: crates/bazed/src/main.rs [hunk 17]
```diff
@@ -1,3 +1,10 @@
...
+use color_eyre::Result;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    bazed_core::app::start("127.0.0.1:6969").await?;
```
Comment (buffet) [inline crates/bazed/src/main.rs:5]: Good port.

File: crates/bazed/src/main.rs [hunk 18]
```diff
@@ -1,3 +1,10 @@
...
+
+    loop {
```
Comment (buffet) [inline crates/bazed/src/main.rs:7]: I will allow this for now, I want this removed by Jan 16th 2023!.

**Review by buffet — CHANGES_REQUESTED** [review]
Over all this is code.

File: crates/bazed/src/main.rs (hunk 5 above)
Comment (buffet) [reply crates/bazed/src/main.rs:2]: Yes, maybe it should be hidden behind a feature flag?

File: crates/bazed-core/src/app.rs [hunk 19]
```diff
@@ -16,7 +16,7 @@ pub struct App {
 impl App {
     #[tracing::instrument(skip(self))]
     async fn load_file(&mut self, path: std::path::PathBuf) -> Result<()> {
```
Comment (buffet) [inline crates/bazed-core/src/app.rs:18]: Should be some open_scratch or similar, probably document the temporary incorrect code.

File: crates/bazed-core/src/document.rs [hunk 20]
```diff
@@ -21,7 +21,17 @@ pub struct Document {
 }
 
 impl Document {
-    pub fn open(path: PathBuf) -> Result<Self> {
+    pub fn open_tmp() -> Result<Self> {
```
Comment (buffet) [inline crates/bazed-core/src/document.rs:24]: I don't like tmp as a name for source-less buffers, but we can discuss that another time.

**Review by buffet — APPROVED** [review]

**Review by gwimm — APPROVED** [review]
"Lmao" Phrase not found.

**Pull request merged**